### PR TITLE
Add tests and error handling to DecodeExtPubKey/DecodeExtKey. Add [[nodiscard]]. 

### PR DESCRIPTION
**Pull request by practicalswift**
Add error handling to `DecodeExtPubKey`/`DecodeExtKey` and make sure return value is checked (via `[[nodiscard]]`).

Prior to this commit:

```
$ cat > test.cpp
…
int main(void) {
  CExtPubKey key = DecodeExtPubKey("foo");
  std::cout << key.nChild << "\n";
}
^D
$ g++ -o test test.cpp
$ ./test
109452364
$ ./test
-1789494196
$ ./test
-1568076724
$ ./test
1483189324
$ ./test
-1168606132
```

### CODE REVIEW

File: src/script/descriptor.cpp [hunk 1]
```diff
@@ -464,9 +464,11 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(const Span<const char>& sp, bool per
             return MakeUnique<ConstPubkeyProvider>(pubkey);
         }
     }
-    CExtKey extkey = DecodeExtKey(str);
-    CExtPubKey extpubkey = DecodeExtPubKey(str);
-    if (!extkey.key.IsValid() && !extpubkey.pubkey.IsValid()) return nullptr;
+    CExtKey extkey;
+    bool validExtKey = DecodeExtKey(str, extkey);
+    CExtPubKey extpubkey;
+    bool validExtPubKey = DecodeExtPubKey(str, extpubkey);
```
Comment (domob1812) [inline src/script/descriptor.cpp]: Nit: This and `validExtKey` above can be const.

File: src/script/descriptor.cpp [hunk 2]
```diff
@@ -464,9 +464,11 @@ std::unique_ptr<PubkeyProvider> ParsePubkey(const Span<const char>& sp, bool per
...
+    if (!(validExtKey && extkey.key.IsValid()) && !(validExtPubKey && extpubkey.pubkey.IsValid())) return nullptr;
```
Comment (domob1812) [inline src/script/descriptor.cpp]: Why do you check `IsValid` here again?  That's already contained in the return value you added.  So either you should make it explicit in `DecodeExtKey`'s contract that the returned key is valid if `true` is returned and remove the check here (preferred for me), or you should not check the key for validity in `DecodeExtKey` and have the callers check it - but not both.

**Review by domob1812 — APPROVED** [review]
utACK 6ade9e8e463ff1843a025ffb7d79fe76870c6342 with a few nits

**Comment by practicalswift** [comment]
@domob1812 Thanks for reviewing. Feedback addressed. Please re-review :-)

**Comment by domob1812** [comment]
Thanks for the fixes, looks good to me - utACK 067409221dde450ee9ba5dd7a545d9640ed7608b.

File: src/test/bip32_tests.cpp [hunk 3]
```diff
@@ -100,11 +100,69 @@ static void RunTest(const TestVector &test) {
...
+        BOOST_CHECK(!DecodeExtPubKey("\r\a" + derive.pub, extpubkeyTmp));
+        BOOST_CHECK(DecodeExtPubKey("\t" + derive.pub, extpubkeyTmp));
+        BOOST_CHECK(!DecodeExtPubKey("\a\t" + derive.pub, extpubkeyTmp));
+        BOOST_CHECK(DecodeExtPubKey(derive.pub + "\n", extpubkeyTmp));
+        BOOST_CHECK(!DecodeExtPubKey(derive.pub + "\a\n", extpubkeyTmp));
+        BOOST_CHECK(DecodeExtPubKey(derive.pub + "\r", extpubkeyTmp));
+        BOOST_CHECK(!DecodeExtPubKey(derive.pub + "\r\a", extpubkeyTmp));
+
+        // Test expected similarities/differences in return values between DecodeExtKey and DecodeExtPubKey
+        BOOST_CHECK(DecodeExtKey(derive.prv, extkeyTmp) != DecodeExtPubKey(derive.prv, extpubkeyTmp));
```
Comment (ryanofsky) [inline src/test/bip32_tests.cpp]: I don't understand the style used in this part of the test. It seems like the test would be both more useful and more readable with:

```c++
BOOST_CHECK(DecodeExtKey(derive.prv, extkeyTmp) != DecodeExtPubKey(derive.prv, extpubkeyTmp));
```
replaced by something like:

```c++
BOOST_CHECK_EQUAL(DecodeExtKey(derive.prv, extkeyTmp, true);
BOOST_CHECK_EQUAL(DecodeExtPubKey(derive.prv, extpubkeyTmp), false);
```

**Review by ryanofsky — COMMENTED** [review]
utACK 067409221dde450ee9ba5dd7a545d9640ed7608b

**Comment by practicalswift** [comment]
AppVeyor build failure seems unrelated? :-)

**Comment by ryanofsky** [comment]
> AppVeyor build failure seems unrelated? :-)

It looks related to lack of visual c++ support for `__attribute__`. You can probably avoid it by checking for `_MSC_VER`. It's probably fine to do this in another PR, though it might be nice to get out of the way now.

Relevant logs:

```
c:\projects\bitcoin\src\key_io.h(21): error C2065: 'warn_unused_result': undeclared identifier [C:\projects\bitcoin\build_msvc\libbitcoin_wallet\libbitcoin_wallet.vcxproj]
c:\projects\bitcoin\src\key_io.h(21): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int [C:\projects\bitcoin\build_msvc\libbitcoin_wallet\libbitcoin_wallet.vcxproj]
c:\projects\bitcoin\src\key_io.h(21): error C2448: '__attribute__': function-style initializer appears to be a function definition [C:\projects\bitcoin\build_msvc\libbitcoin_wallet\libbitcoin_wallet.vcxproj]
c:\projects\bitcoin\src\key_io.h(24): error C2065: 'warn_unused_result': undeclared identifier [C:\projects\bitcoin\build_msvc\libbitcoin_wallet\libbitcoin_wallet.vcxproj]
c:\projects\bitcoin\src\key_io.h(24): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int [C:\projects\bitcoin\build_msvc\libbitcoin_wallet\libbitcoin_wallet.vcxproj]
c:\projects\bitcoin\src\key_io.h(24): error C2374: '__attribute__': redefinition; multiple initialization [C:\projects\bitcoin\build_msvc\libbitcoin_wallet\libbitcoin_wallet.vcxproj]
  c:\projects\bitcoin\src\key_io.h(21): note: see declaration of '__attribute__'
c:\projects\bitcoin\src\key_io.h(24): error C2448: '__attribute__': function-style initializer appears to be a function definition [C:\projects\bitcoin\build_msvc\libbitcoin_wallet\libbitcoin_wallet.vcxproj]
```

**Review by ryanofsky — COMMENTED** [review]
utACK a7db4f76e3700f17f3d627eee73b338f86b9c254. Only change since previous review was suggested test cleanup (Thanks!)

**Comment by practicalswift** [comment]
@ryanofsky Updated to work with Visual C++ as well. Please re-review :-)

File: src/attributes.h [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+#ifndef BITCOIN_ATTRIBUTES_H
+#define BITCOIN_ATTRIBUTES_H
+
+#if defined(__has_cpp_attribute) && __has_cpp_attribute(nodiscard)
+#  define NODISCARD [[nodiscard]]
+#elif defined(_MSC_VER) && _MSC_VER >= 1700
+#  define NODISCARD _Check_return_
+#else
+#  define NODISCARD __attribute__((warn_unused_result))
```
Comment (luke-jr) [inline src/attributes.h]: This isn't guaranteed to be valid AFAIK?

File: src/key_io.cpp [hunk 5]
```diff
@@ -159,17 +159,17 @@ std::string EncodeSecret(const CKey& key)
     return ret;
 }
 
-CExtPubKey DecodeExtPubKey(const std::string& str)
+bool DecodeExtPubKey(const std::string& str, CExtPubKey& extpubkey)
```
Comment (luke-jr) [inline src/key_io.cpp]: I think the current API is better. Just check `.IsValid()` in the caller...

File: src/attributes.h (hunk 4 above)
Comment (practicalswift) [reply src/attributes.h]: Do you have any suggestion on a more proper way to do it?

FWIW this is how Chromium does it :-)

```
#undef WARN_UNUSED_RESULT
#if defined(COMPILER_GCC) || defined(__clang__)
#define WARN_UNUSED_RESULT __attribute__((warn_unused_result))
#else
#define WARN_UNUSED_RESULT
#endif
```

**Review by ryanofsky — APPROVED** [review]
utACK 273a622e058c8eafaac56393e0a3bf7d8868ab4f. Only change since previous review is adding NODISCARD for visual c

**Pull request closed without merging**